pkg/iotool/timeoutwriter.go

```diff
@@ -0,0 +1,56 @@
+package iotool
+
+import (
+	"fmt"
+	"io"
+	"sync"
+	"time"
+)
+
+/*
+	TimeoutWriter accepts input until expiration, then returns errors.
+
+	An error will be written to the wrapped writer unless `Finished` is called
+	before the timeout.
+*/
+type TimeoutWriter struct {
+	w    io.Writer
+	mtx  sync.Mutex
+	done bool
+}
+
+var TimeoutErr = fmt.Errorf("writer closed due to timeout")
+
+func NewTimeoutWriter(w io.Writer, patience time.Duration) *TimeoutWriter {
+	return NewTimeoutWriterFromChan(w, time.After(patience))
+}
+
+func NewTimeoutWriterFromChan(w io.Writer, done <-chan time.Time) *TimeoutWriter {
+	t := &TimeoutWriter{w: w}
+	go func() {
+		<-done
+		t.mtx.Lock()
+		defer t.mtx.Unlock()
+		t.done = true
+		if w != nil {
+			fmt.Fprintf(w, "%s\n", TimeoutErr)
```

Comment:
Could, though without parametric polymorphism the constructor names are going to get hilarious.

In the current use case, it doesn't really matter, because the uses are actually serial anyway.  The only reason it raced was in that when the timeout caused the main thread to move on, nothing stopped the slow writers.